### ISSUE TITLE
Best practice for adding a column for `__filename` without adding a duplicate field?

### ISSUE THREAD

**Issue by grisaitis** [apache/arrow#34554]
#### my problem
i have a `Dataset` and want to add the `__filename` field when creating a `Table` from this dataset. I've tried a few things:

#### attempt 1: use `dataset.scanner`

```python
columns = dataset.schema.names + ["__filename"]  # ["__last_in_fragment"]
scanner = dataset.scanner(columns=columns)
my_table = scanner.to_table()
```

but with this, i get a duplicate column error when i try to access it with duckdb:
```python
con = duckdb.connect()
con.execute("SELECT * FROM my_table limit 5").fetch_df()
```
```
duckdb.InvalidInputException: Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
Error: Invalid Error: ArrowInvalid: Multiple matches for FieldRef.Name(__filename)
```

#### attempt 2: use a dict of columns
i have also tried creating the new column by specifying columns as expressions. but, `ds.field("__filename")` is not recognized:

```python
columns = {x: ds.field(x) for x in dataset.schema.names}
columns["filename"] = ds.field("__filename")
scanner = dataset.scanner(columns=columns)
```

```
Traceback (most recent call last)
----> scanner = dataset.scanner(columns=columns)
/opt/conda/envs/myenv/lib/python3.11/site-packages/pyarrow/_dataset.pyx:336, in pyarrow._dataset.Dataset.scanner()
/opt/conda/envs/myenv/lib/python3.11/site-packages/pyarrow/_dataset.pyx:2576, in pyarrow._dataset.Scanner.from_dataset()
/opt/conda/envs/myenv/lib/python3.11/site-packages/pyarrow/_dataset.pyx:2484, in pyarrow._dataset.Scanner._make_scan_options()
/opt/conda/envs/myenv/lib/python3.11/site-packages/pyarrow/_dataset.pyx:2381, in pyarrow._dataset._populate_builder()
/opt/conda/envs/myenv/lib/python3.11/site-packages/pyarrow/error.pxi:100, in pyarrow.lib.check_status()
ArrowInvalid: No match for FieldRef.Name(__filename) in <schema...>
```

#### tl;dr
what is the best practice for adding fields like `__filename` to a table while avoiding duplicate column errors?

#### related issues

this might be related to
- https://github.com/apache/arrow/issues/24407
- https://github.com/dask/dask/issues/9251

#### versions

```
Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.28
Python 3.11.0 | packaged by conda-forge (main, Jan 14 2023, 12:27:40) [GCC 11.3.0]
pyarrow 11.0.0
```

### Component(s)

Python

**Comment by jwolosiuk** [comment]
Hi, is there any update on this, or any workaround? I got the same problem with version 13.0.0

**Comment by dhirschfeld** [comment]
Also running into this 😔

**Comment by dhirschfeld** [comment]
Seems to work, though I'm not sure how efficient it is :/
```python
def rename_columns(schema: pa.Schema, **name_mapping) -> pa.RecordBatch:
    return pa.schema(
        field.with_name(name_mapping.get(field.name, field.name))
        for field in schema
    )


def append_filename(ds: pa.dataset.Dataset) -> pa.dataset.Dataset:
    scanner = ds.scanner(columns=ds.schema.names + ['__filename'])
    schema = rename_columns(scanner.projected_schema, __filename='filename')
    batches = (
        pa.RecordBatch.from_arrays(batch.columns, schema=schema)
        for batch in scanner.to_batches()
    )
    return pa.dataset.dataset(list(batches))   
```

**Comment by dhirschfeld** [comment]
> Seems to work, though I'm not sure how efficient it is :/

Needless to say, it would be good to have this natively (and efficiently!) supported by `pyarrow` itself.

One workaround would be to allow columns to be renamed in the scanner, similarly to how I'm doing it above but doing it on they fly rather than reiterating and mutating the dataset after the fact.

**Comment by StashOfCode** [comment]
What's the status of this? I can't get a "__filename" from a dataset without explicitly listing a schema first which kinda ruins the whole point of schema detection, especially if there's a variable number of columns or column names were changed/are messed up.
